Working pixel rendering on host using wxWidgets.

lawik opened the pull request.

lawik committed:
Working pixel rendering on host using wxWidgets.

nyaray commented on .gitignore:
```diff
@@ -16,4 +16,5 @@
-upload.sh
\ No newline at end of file
+upload.sh
+.elixir_ls
```
nice.

nyaray commented on draw_some_pixels.esx:
```diff
@@ -0,0 +1,77 @@
+defmodule Script do
```
file ending should be `exs`

nyaray commented on draw_some_pixels.esx:
```diff
@@ -0,0 +1,77 @@
+                  true ->
+                    :white
+                end
+            end
```
Anti-pattern alert re: over use/nesting of `cond`s (which I'm only commenting on as I've noticed them as a pattern) and, sort of, two-case `cond`s, which should be `if-else`s.

You should flatten these deep and verbose structures into `case`s, along the lines of:

``` elixir
# naming these probably isn't necessary, though 🤷‍♂ 
x_big = x > width / 2
y_big = y > height / 2

color =
  case {x_big, y_big} do
    {true, true} -> :accent
    {true, false} when rem(x, 2) == 0 -> :white
    {true, false} -> :black
    {false, true} -> :black
    {false, false} -> :white
  end
```

lawik replied:
This is a copy paste of the testrun stuff you had the same thoughts on.
But that looks clean. Should try it.

nyaray replied:
Don't know which to judge you most on, the copypasting or not wanting to improve the copypaste 😜

lawik replied:
It worked, so it was perfect ;)

nyaray commented on lib/host/canvas.ex:
```diff
@@ -0,0 +1,101 @@
+        :red -> {255, 0, 0}
+        :yellow -> {255, 255, 0}
+        _ -> {255, 0, 0}
+      end
```
Unknown colors should be obviously wrong, maybe magenta would do?

lawik replied:
0,255,0 supergreen :)

nyaray commented on lib/host/canvas.ex:
```diff
@@ -0,0 +1,101 @@
+
+    for y <- 0..(height - 1),
+        x <- 0..(width - 1),
+        do: draw_pixel(dc, x, y, brushes, pixels[{x, y}])
```
much neater than those nested `reduce`s!

nyaray commented on lib/host/canvas.ex:
```diff
@@ -0,0 +1,101 @@
+defmodule Inky.Host.Canvas do
```
Adding the host subdirectory feels like an unnecessarily hierarchical choice. I suggest we keep things flat until there are many of the "it"s we want to group.

lawik replied:
I would rather silo the host code which would include HostCommands or whatever the module might be called.  Mostly because it is irrelevant to understanding most of the library, it is off the beaten path, it is an extra thing. Honestly it might be relevant to make it a separate lib to avoid it getting compiled to nerves builds and stuff.
So it has a reason. I want people to be able to grasp the important code of the project at a glance and ignore this.
Does that make sense as a use for a subdirectory for you?

nyaray replied:
Hm... those are good points about the file layout saying guiding the reader's eyes, but I think your idea of making it a separate lib would be even better, then it can be a dev-dependency, just as you said!

Side-note: `HostCommands` sounds like a good name since it's the commands specific to the host-target :)

lawik replied:
Yeah, I figure since we have TestIO, RpiIO and if I had done the IO bit for this it would have been HostIO. So I imagine something similar will be done for the commands?

nyaray replied:
Pretty much, yeah :)

nyaray reviewed (approved):
Looks neat overall; we'll unify it once the `Commands` behaviour is in `master`

lawik committed:
Added .tool-versions to .gitignore.
Fixed naming of file draw_some_pixels.exs.
Cleaned up some pixel-builder-code from review.
Changed the behavior for bad colors in the host canvas.

lawik commented:
Closing this to move the functionality to a separate lib.

lawik commented:
See https://github.com/pappersverk/inky_host_dev
